sky/packages/sky/lib/src/fn3/framework.dart

```diff
@@ -712,11 +719,11 @@ abstract class BuildableElement<T extends Widget> extends Element<T> {
   /// components dirty during event handlers before the frame begins, not during
   /// the build itself.
   void markNeedsBuild() {
-    assert(!_debugStateLocked);
     assert(_debugLifecycleState == _ElementLifecycle.mounted);
     if (_dirty)
       return;
     _dirty = true;
+    assert(!_debugInBuildScope);
```

Comment:
> Do we catch setState() being invoked on an already-dirty child?

We don't.

> If not, that would fail when the child is not dirty but not fail when it's already dirty, which would lead to some really hard-to-nail-down bugs.

That's true.  I could track the whether a given component is building itself and use that as the white list.